File: lektionen/javascript/Kata/k5_common_denominators.js
function convertFrac(lst){
  var num =[];
  var den = []; 
  var lcd;
  var result = [];
  var ans = "";

  for (let element of lst) {
    num.push(element[0]);
    den.push(element[1]);
  }
//If the greatest denominator from the array is also the LCD
  if (den.every(x => Math.max(...den) % x == 0)) {
    lcd = Math.max(...den);
    for (let i = 0; i < den.length; i++) {
      result.push([]);
      result[i].push(lcd / den[i] * num[i]);
      result[i].push(lcd);
    }
    for (let i = 0; i < result.length; i++) {
      ans += "(" + result[i][0].toString() + "," + result[i][1].toString() + ")"
    }
    return ans;
  }
//If not, start from the product of the denominators and try to guess 'g'
//by taking steps of multiples of the greatest denominator until it reaches
//the product of the denominators divided by the greatest denominator, if 
//any multiples are found, then the lowest stays the product of the denominators
 
  else {
    lcd = den.reduce((a, b) => a * b);
    var g = [];

    for (let i = 2; i < lcd / Math.max(...den); i++) {
      g.push(Math.max(...den) * i);
    }

    for (let i = g.length - 1; i >= 0; i--) {
      if (den.every(x =>  g[i] % x == 0)){
        lcd = g[i];
      }
    }

    for (let i = 0; i < den.length; i++) {
      result.push([]);
      result[i].push(lcd / den[i] * num[i]);
      result[i].push(lcd);
    }

    for (let i = 0; i < result.length; i++) {
      ans += "(" + result[i][0].toString() + "," + result[i][1].toString() + ")"
    }

    return ans;
  }
}

File: lektionen/python/Kata/k5_common_denominators.py
def convertFracts(lst:list) -> list:
  if len(lst) == 0: return []
  def isCommonDivisor(n:int,denoms:list) -> bool:
    """Checks whether an integer is a common divisor among a list of denominators"""
    c = 0
    for i in denoms: 
      if n % i == 0:
        c+=1
    if c == len(denoms):
      return True
    return False
    
  #check within denominators
  denoms = [x[1] for x in lst]
  hd = max(denoms)
  if isCommonDivisor(hd, denoms):
    lcd = hd
  #check multiples of highest denominator:
  else:
    um = 1
    for i in denoms:
      if i != hd:
        um*=i 
    for j in range(um*hd, 1, -hd):
      if isCommonDivisor(j, denoms):
        lcd = j

  return [[lcd//x[1]*x[0],lcd] for x in lst]

print(convertFracts([[1, 3], [1, 6], [1, 15]]))